flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/PipelineStage.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.pipeline;
+
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.api.misc.param.WithParams;
+
+/**
+ * Abstract class for a stage in a pipeline, either an [[Estimator]] or a [[Transformer]].
+ *
+ * <p>Each pipeline stage is with parameters, and requires a public empty constructor for
+ * restoration in Pipeline. It implement the {@link org.apache.flink.ml.api.core.PipelineStage}
+ * and hold a {@link Params} as its member, thus the subclasses could do not care about
+ * {@link WithParams#getParams()}
+ *
+ * @param <S> The class type of the {@link PipelineStage} implementation itself, used by {@link
+ *            org.apache.flink.ml.api.misc.param.WithParams} and Cloneable.
+ */
+public abstract class PipelineStage<S extends PipelineStage <S>>
```

Comment:
Same here, `PipelineStageBase`? This class does not implement the `org.apache.flink.ml.api.core.PipelineStage`, is this intended?

Reply:
I am not sure this class is a must. I think each type of pipeline stage individually can set its own base class. see: https://stackoverflow.com/a/3599379/11332462

Reply:
> Same here, `PipelineStageBase`? This class does not implement the `org.apache.flink.ml.api.core.PipelineStage`, is this intended?

> I am not sure this class is a must. I think each type of pipeline stage individually can set its own base class. see: https://stackoverflow.com/a/3599379/11332462

It has been renamed and PipelineStage is the standard concept in pipeline and if it implements the `WithParams` , the subclasses could do not care about `WithParams`